### PR TITLE
Add user accounts with Google OAuth2.0

### PR DESCRIPTION
**Pull request by kgang**
* Use Google OIDC ID Tokens for Auth
* Create auth middlewares for API and SaaS
* Cleaned up dockerfiles
* Added env vars and .env.example
* Created models for `User`, `AuthState`, and `AuthToken`

### CODE REVIEW

**Comment by kgang** [comment]
I made edits to **all** the route in the API, but I think I'm better off with a middle ware the way I have it set up

**Comment by kgang** [comment]
Reading [this ](https://www.notion.so/OAuth-Identity-113b88e224dd8015b241dc2dee5b8f6e)is a good starting point.

**Comment by kgang** [comment]
I'm taking my final pass on this now, then I'll set this ready to review

I have to create the staging oauth client id and staging .env files too

File: concrete/utils.py [hunk 1]
```diff
@@ -1,9 +1,24 @@
 """AI generated"""
```
Comment (MichaelDeng03) [inline concrete/utils.py:1]: Why is this file here?

File: docs/developer-guide.md [hunk 2]
```diff
@@ -120,3 +120,14 @@ To create a new migration script, run `alembic revision --autogenerate -m 'migra
...
+## Web Dev
+
+### Allow local subdomains
+By default, auth will be enabled on webapps even when run locally.
+This requires an auth service to be run, which is hosted at a sub-domain in staging/prod.
+To mirror this setup locally add the following lines to the bottom of `/etc/hosts`
+
+```
+127.0.0.1 abop.bot auth.abop.bot
+```
```
Comment (MichaelDeng03) [inline docs/developer-guide.md:133]:
```suggestion
```
### Auth
Auth Flow: 

Google OIDC Client Secret + User Google Login -> Google, returning an Auth Code
Auth Code -> Google, returning ID Token and Refresh Token
ID Token can be verified using Google's public keys, and represent a users identity

API Auth: Google's ID Token will be used as an Auth Token for Abops API, and the client will be responsible for not losing it.
SaaS Auth: Google's ID Token will be used as Auth Token, but it will be stored in cookies. CORS enables sharing of cookies between auth.abop.ai -> abop.ai, then SaaS can read cookies from abop.ai

We use the OAuth2 standard protocol enabled through google's SDKs to identify and authenticate users.

1. The user initiates login with AO
2. The user is redirected to google to complete authentication
  a. If it is the user's first time, they will have to authorize Google
  to provide their user data (name, email, profile) to AO
  b. Only URLs designated in the OAuth client on GCloud will be allowed to make this redirect with some client ID
    - This is how Google knows that they aren't fielding a random request
1. Google returns the user to a callback url on AO with a special authorization code
  a. This redirect url is also hardcoded in GCloud
1. AO uses the authorization code in the auth_callback with the Google auth sdk to receive a refresh token for the user
2. The user has now been authenticated
  a. AO stores the refresh token and creates an account if needed
1. The user is redirected away again from the auth_callback endpoint to the landing page
or the page they were originally trying to access
```

File: flake.txt
Comment (MichaelDeng03) [inline flake.txt:1]: rm?

File: webapp/api/server.py [hunk 3]
```diff
@@ -18,15 +24,39 @@
...
     OrchestratorUpdate,
 )
+from concrete.webutils import AuthMiddleware
 
 from .models import CommonReadParameters
 
+dotenv.load_dotenv(override=True)
+
+UNAUTHENTICATED_PATHS = {'/docs', '/redoc', '/openapi.json', '/favicon.ico'}
+
```
Comment (MichaelDeng03) [inline webapp/api/server.py:34]: dym writing up some more notes on how middleware is sharing tokens?

**Review by MichaelDeng03 — COMMENTED** [review]
Couldn't get a thorough review in unfortunately

File: concrete/utils.py (hunk 1 above)
Comment (MichaelDeng03) [reply concrete/utils.py:1]: oh nvm lol

File: webapp/api/server.py (hunk 3 above)
Comment (kgang) [reply webapp/api/server.py:34]: added to auth server.py

Comment (MichaelDeng03) [reply webapp/api/server.py:34]: looks like it's in api/server.py [](edited-with-Axolo)

**Review by MichaelDeng03 — APPROVED** [review]
